browser/importer/extensions_import_helpers.cc

```diff
@@ -5,92 +5,330 @@
+#include "chrome/browser/extensions/extension_service.h"
+#include "chrome/browser/extensions/webstore_install_with_prompt.h"
+#include "chrome/browser/profiles/profile.h"
 #include "components/value_store/value_store.h"
 #include "components/value_store/value_store_factory.h"
 #include "components/value_store/value_store_factory_impl.h"
 #include "extensions/browser/api/storage/value_store_util.h"
 #include "extensions/browser/extension_file_task_runner.h"
+#include "extensions/browser/extension_registry.h"
+#include "extensions/browser/extension_system.h"
 #include "extensions/common/constants.h"
 
-namespace brave {
-namespace {
-using ExtensionStorageMap =
-    base::flat_map<std::string, std::unique_ptr<value_store::ValueStore>>;
+namespace extensions_import {
 
-std::vector<std::string> GetChromeExtensionsSettingsList(
+// Silent installer via webstore w/o any prompt or bubble.
+class WebstoreInstallerForImporting
+    : public extensions::WebstoreInstallWithPrompt {
+ public:
+  using WebstoreInstallWithPrompt::AbortInstall;
+  using WebstoreInstallWithPrompt::WebstoreInstallWithPrompt;
+
+ private:
+  ~WebstoreInstallerForImporting() override = default;
+
+  std::unique_ptr<ExtensionInstallPrompt::Prompt> CreateInstallPrompt()
+      const override {
+    return nullptr;
+  }
+  bool ShouldShowPostInstallUI() const override { return false; }
+};
+
+base::expected<std::vector<ImportingExtension>, bool> GetExtensionsList(
     const base::FilePath& source_profile,
-    std::vector<std::string> extensions_ids) {
-  auto store_factory_source =
+    const base::FilePath& target_profile) {
+  std::vector<ImportingExtension> result;
+  const auto extensions = GetImportableChromeExtensionsList(source_profile);
+  if (!extensions || extensions->empty()) {
+    return base::ok(std::move(result));
+  }
+
+  auto source_store_factory =
       base::MakeRefCounted<value_store::ValueStoreFactoryImpl>(source_profile);
-  std::vector<std::string> result;
-  for (const auto& id : extensions_ids) {
-    if (!store_factory_source->HasValueStore(
+
+  for (const auto& extension_id : *extensions) {
+    const auto value_store =
+        base::FilePath(extensions::kLocalExtensionSettingsDirectoryName)
+            .AppendASCII(extension_id);
+
+    bool has_local_settings = false;
+
+    if (source_store_factory->HasValueStore(
             base::FilePath(extensions::kLocalExtensionSettingsDirectoryName)
-                .AppendASCII(id)))
-      continue;
-    result.push_back(id);
+                .AppendASCII(extension_id))) {
+      auto store = extensions::value_store_util::CreateSettingsStore(
+          extensions::settings_namespace::LOCAL,
+          extensions::value_store_util::ModelType::EXTENSION, extension_id,
+          source_store_factory);
+      const auto settings = store->Get();
+      if (!settings.status().ok()) {
+        if (!settings.status().IsCorrupted()) {
+          return base::unexpected(false);
```

Comment:
should this really abort the import for all extensions or just ignore a single one?

Reply:
We don't have any ui to show what is imported and what isn't, so if preparation for any extension has failed we decide to not start the import process while user doesn't close the source browser.